subsys/bluetooth/services/cts.c

```diff
@@ -0,0 +1,192 @@
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(cts, CONFIG_BT_CTS_LOG_LEVEL);
+
+#define BT_CTS_ATT_ERR_VALUES_IGNORED 0x80
+#define BT_CTS_FRACTION_256_MAX_VALUE 255
+
+static const struct bt_cts_cb *_cb;
+
+#if CONFIG_BT_CTS_HELPER_API
+
+#include <time.h>
+#include <zephyr/sys/timeutil.h>
+
+int bt_cts_time_to_unix_ms(const struct bt_cts_time_format *ct_time, int64_t *unix_ms)
+{
+	struct tm date_time;
+	/* fill date time structure */
+	date_time.tm_year = sys_le16_to_cpu(ct_time->year); /* year (little endian) */
+	date_time.tm_year -= TIME_UTILS_BASE_YEAR;
+	date_time.tm_mon = ct_time->mon - 1;   /* month start from 1, but need from 0 */
+	date_time.tm_mday = ct_time->mday;     /* day of month */
+	date_time.tm_hour = ct_time->hours;    /* hours of day */
+	date_time.tm_min = ct_time->min;       /* minute of hour */
+	date_time.tm_sec = ct_time->sec;       /* seconds of minute */
+	date_time.tm_wday = ct_time->wday % 7; /* for sundays convert to 0, else keep same */
+
+	LOG_DBG("CTS Write Time: %d/%d/%d %d:%d:%d", date_time.tm_year, date_time.tm_mon,
+		date_time.tm_mday, date_time.tm_hour, date_time.tm_min, date_time.tm_sec);
+	/* get unit timestamp from datetime */
+	(*unix_ms) = timeutil_timegm64(&date_time);
+	if ((*unix_ms) == ((time_t)-1)) {
+		return -EOVERFLOW;
+	}
+	LOG_DBG("CTS Write Unix: %lld", (*unix_ms));
+	(*unix_ms) *= MSEC_PER_SEC;
+	/* add fraction 256 part*/
+	(*unix_ms) += ((ct_time->fractions256 * MSEC_PER_SEC) / BT_CTS_FRACTION_256_MAX_VALUE);
+
+	return 0;
+}
+
+int bt_cts_time_from_unix_ms(struct bt_cts_time_format *ct_time, int64_t unix_ms)
+{
+	struct tm date_time;
+	time_t unix_ts = unix_ms / MSEC_PER_SEC;
+
+	/* 'Fractions 256 part of 'Exact Time 256' */
+	unix_ms %= MSEC_PER_SEC;
+	unix_ms *= BT_CTS_FRACTION_256_MAX_VALUE;
+	unix_ms /= MSEC_PER_SEC;
+	ct_time->fractions256 = unix_ms;
+
+	/* convert unix_ts to */
+	LOG_DBG("CTS Read Unix: %lld", unix_ts);
+	/* generate date time from unix timestamp */
+	if (gmtime_r(&unix_ts, &date_time) == NULL) {
+		return -EOVERFLOW;
+	}
+	date_time.tm_year += TIME_UTILS_BASE_YEAR;
+
+	LOG_DBG("CTS Read Time: %d/%d/%d %d:%d:%d", date_time.tm_year, date_time.tm_mon,
+		date_time.tm_mday, date_time.tm_hour, date_time.tm_min, date_time.tm_sec);
+
+	/* 'Exact Time 256' contains 'Day Date Time' which contains
+	 * 'Date Time' - characteristic contains fields for:
+	 * year, month, day, hours, minutes and seconds.
+	 */
+	ct_time->year = sys_cpu_to_le16(date_time.tm_year);
+	ct_time->mon = date_time.tm_mon + 1; /* months starting from 1 */
+	ct_time->mday = date_time.tm_mday;   /* Day of month */
+	ct_time->hours = date_time.tm_hour;  /* hours */
+	ct_time->min = date_time.tm_min;     /* minutes */
+	ct_time->sec = date_time.tm_sec;     /* seconds */
+	/* day of week starting from 1-monday, 7-sunday */
+	ct_time->wday = date_time.tm_wday;
+	if (ct_time->wday == 0) {
+		ct_time->wday = 7; /* sunday is represented as 7 */
+	}
+	return 0;
+}
+
+#endif
```

Comment:
Especially for longer #ifdef..#endif blocks its good practice to put a comment reminding which condition terminates here, i.e. something like:
```c
#endif /* CONFIG_BT_CTS_HELPER_API */
```